firestore.rules

```diff
@@ -41,37 +41,92 @@ service cloud.firestore {
       return hasUserId() && hasRole("teacher");
     }
 
+    // uid of submitted document must match user's platform_user_id
     function requestMatchUserId() {
       return isAuthed() &&
         string(request.auth.token.platform_user_id) == request.resource.data.uid;
     }
 
-    function docMatchUserId() {
+    // user's platform_user_id must be in submitted document's list of teachers
+    function requestMatchTeachers() {
+      return isAuthed() &&
+        string(request.auth.token.platform_user_id) in request.resource.data.teachers;
+    }
+
+    // uid of requested document must match user's platform_user_id
+    function resourceMatchUserId() {
       return isAuthed() &&
         string(request.auth.token.platform_user_id) == resource.data.uid;
     }
 
+    // user's platform_user_id must be in requested document's list of teachers
+    function resourceMatchTeacher() {
+      return isAuthed() &&
+        string(request.auth.token.platform_user_id) in resource.data.teachers;
+    }
+
+    // user's class_hash must be in submitted document's list of classes
     function classInRequestClasses() {
       return isAuthed() && request.auth.token.class_hash in request.resource.data.classes;
     }
 
+    // user's class_hash must be submitted document's context_id
     function classIsRequestContextId() {
       return isAuthed() && request.auth.token.class_hash == request.resource.data.context_id;
     }
 
-    function hasSupportContentAndType() {
-      return (request.resource.data.content != null) && (request.resource.data.type == "supportPublication");
+    // user's class_hash must be in requested document's list of classes
+    function classInResourceClasses() {
+      return isAuthed() && request.auth.token.class_hash in resource.data.classes;
+    }
+
+    function isValidDocumentCreateRequest() {
+      return requestMatchTeachers() && classIsRequestContextId() &&
+              request.resource.data.keys().hasAll(["uid", "type", "key", "createdAt"]);
+    }
+
+    function preservesReadOnlyDocumentFields() {
+      let readOnlyFieldsSet = ["uid", "type", "key", "createdAt"].toSet();
+      let affectedFieldsSet = debug(debug(request.resource.data).diff(resource.data).affectedKeys());
+      return !affectedFieldsSet.hasAny(readOnlyFieldsSet);
+    }
+
+    function isValidDocumentUpdateRequest() {
+      return resourceMatchTeacher() && preservesReadOnlyDocumentFields();
     }
 
-    function isValidSupportRequest() {
+    function isValidCommentCreateRequest() {
+      return requestMatchUserId() &&
+              request.resource.data.keys().hasAll(["name", "createdAt", "content"]);
+    }
+
+    function preservesReadOnlyCommentFields() {
+      let readOnlyFieldsSet = ["uid", "createdAt", "tileId"].toSet();
+      let affectedFieldsSet = debug(debug(request.resource.data).diff(resource.data).affectedKeys());
```

Comment:
This is a nice approach. Do you mean to leave these `debug` functions in here?

Reply:
Nope -- good catch. I'll remove them and then merge the PR.